R/dplyr.R

```diff
@@ -261,17 +261,36 @@ prepare_join <- function(x, y, by, selected, suffix, copy, disambiguate = TRUE)
   if (quo_is_null(select_quo)) {
     selected <- tidyselect::vars_select(all_cols_y, everything())
   } else {
+    # FIXME: should we complain, if the user wishes to rename two or more columns to the same name?
     selected <- tidyselect::vars_select(all_cols_y, !!select_quo)
   }
 
   if (is_null(by)) {
+    rename_rhs_by <- TRUE
     by <- get_by(x, x_orig_name, y_name)
-    if (!any(selected == by)) abort_need_to_select_rhs_by(y_name, unname(by))
-
     if (!all(names(by) %in% get_tracked_keys(x))) abort_fk_not_tracked(x_orig_name, y_name)
+    # here we need to check if the RHS `by`-column taken directly from the `dm` key relations
+    # is part of the values of `selected` (even if it is renamed)
+    if (has_length(setdiff(by, selected))) {
```

Comment:
Can you extract `setdiff(by, selected)` to a variable? (Can we make the behaviors in the two cases more similar?)

Reply:
not sure: in the once case we need `setdiff(by, selected)` and in the case we need `setdiff(by, names(selected))` (at least if we stick to how I understood the situation in the first place)